libgnucash/engine/Transaction.c

```diff
@@ -2547,20 +2556,39 @@ xaccTransRetDateDue(const Transaction *trans)
 char
 xaccTransGetTxnType (const Transaction *trans)
 {
-    const char *s = NULL;
-    GValue v = G_VALUE_INIT;
-    char ret = TXN_TYPE_NONE;
-
     if (!trans) return TXN_TYPE_NONE;
-    qof_instance_get_kvp (QOF_INSTANCE (trans), &v, 1, TRANS_TXN_TYPE_KVP);
-    if (G_VALUE_HOLDS_STRING (&v))
+
+    if (trans->txn_type == TXN_TYPE_UNCACHED)
     {
-         s = g_value_get_string (&v);
-         if (s && strlen (s) == 1)
-             ret = s[0];
+        Transaction *txn = (Transaction*) trans;
+        gboolean all_splits_have_lot = TRUE;
+        txn->txn_type = TXN_TYPE_NONE;
+        for (GList *n = xaccTransGetSplitList (txn); n; n = g_list_next (n))
+        {
+            GNCLot *lot = xaccSplitGetLot (n->data);
+            if (!lot)
+                all_splits_have_lot = FALSE;
+            else if (txn->txn_type == TXN_TYPE_NONE)
+            {
+                GncInvoice *invoice = gncInvoiceGetInvoiceFromLot (lot);
+                GncOwner owner;
+                if (invoice)
+                {
+                    txn->txn_type = (txn == gncInvoiceGetPostedTxn (invoice))
+                        ? TXN_TYPE_INVOICE : TXN_TYPE_PAYMENT;
+                    DEBUG ("%p txn type is %c", trans, txn->txn_type);
+                }
+                else if (gncOwnerGetOwnerFromLot (lot, &owner))
+                {
+                    txn->txn_type = TXN_TYPE_PAYMENT;
+                    DEBUG ("%p txn type is unattached payment", trans);
+                }
+            }
+        }
+        if (all_splits_have_lot && (txn->txn_type == TXN_TYPE_PAYMENT))
```

Comment:
This will break if there are any extra splits in the transaction. For example some users have mentioned that they insert random zero-value splits for tagging or for adding comments.  To handle this you could do
```
unsigned int count_lots = 0;
/* ... */
 for (GList *n = xaccTransGetSplitList (txn); n; n = g_list_next (n))
  {

      if (!lot)
         continue;
      if (txn->txn_biz_type !=  TXN_BIZ_TYPE_NONE)
      {
           ++count_lots;
           continue;
      }
     /* Do the type heurisitic */
}
if (count_lots > 1 && (txn->txn_type == TXN_TYPE_PAYMENT))
   txn->txn_type = TXN_BIZ_TYPE_LINK
```

But the comment for `TXN_TYPE_LINK` says "Transaction is a link between (invoice and payment) lots". This test  assigns it if the first split with a lot resolves to a payment and ignores the other splits with lots. What if the first split encountered is an invoice split? What if there are two payment splits, or more than two splits with lots?

Reply:
IIUC `TXN_TYPE_LINK` and `TXN_TYPE_INVOICE` have always been strictly Read-only. @gjanssens could confirm. Wouldn't your code fail when a payment is applied to clear 2 invoices?

Reply:
Wouldn't your code fail when a payment is applied to clear 2 invoices?

Yes, and so will yours.

> IIUC TXN_TYPE_LINK and TXN_TYPE_INVOICE have always been strictly Read-only.

Do you mean that splits with those flags set are read only? How is that significant here, where you're assigning the flags?

Reply:
> > IIUC TXN_TYPE_LINK and TXN_TYPE_INVOICE have always been strictly Read-only.
> 
> Do you mean that splits with those flags set are read only? How is that significant here, where you're assigning the flags?

It would mean Invoice-posting transactions, and Lot-link Payments should never be modifiable by the user to add zero-amount splits to an arbitrary account. However @gjanssens had fixed some lot-link proliferation bugs so I'm not sure my assumption stands.

Reply:
TXN_TYPE_LINK transactions are writable. Only TXN_TYPE_INVOICE transactions are read-only.

Reply:
Ok last commit(s) amended: if we conclude TXN_TYPE_PAYMENT, and there are no non-APAR zero-amount splits, then it must be a TXN_TYPE_LINK.